tpu-client-next/src/leader_updater.rs

```diff
@@ -0,0 +1,117 @@
+    solana_rpc_client::nonblocking::rpc_client::RpcClient,
+    solana_tpu_client::nonblocking::tpu_client::LeaderTpuService,
+    std::{
+        fmt,
+        net::SocketAddr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+    },
+};
+
+/// [`LeaderUpdater`] trait abstracts out functionality required for the
+/// [`ConnectionWorkersScheduler`](crate::network::ConnectionWorkersScheduler)
+/// to identify next leaders to send transactions to.
+#[async_trait]
+pub trait LeaderUpdater: Send {
+    fn get_leaders(&self) -> Vec<SocketAddr>;
+    async fn stop_and_join(&mut self);
+}
+
+pub struct LeaderUpdaterError;
+
+impl fmt::Display for LeaderUpdaterError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Leader updater encountered an error")
+    }
+}
+
+impl fmt::Debug for LeaderUpdaterError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "LeaderUpdaterError")
+    }
+}
+
+/// Creates a [`LeaderUpdater`] based on the configuration provided by the
+/// caller. If a pinned address is provided, it will return a
+/// `PinnedLeaderUpdater` that always returns the provided address instead of
+/// checking leader schedule. Otherwise, it creates a `LeaderUpdaterService`
+/// which dynamically updates the leaders by connecting to the network via the
+/// [`LeaderTpuService`].
+pub async fn create_leader_updater(
+    rpc_client: Arc<RpcClient>,
+    websocket_url: String,
+    lookahead_slots: u64,
+    pinned_address: Option<SocketAddr>,
+) -> Result<Box<dyn LeaderUpdater>, LeaderUpdaterError> {
+    if let Some(pinned_address) = pinned_address {
+        return Ok(Box::new(PinnedLeaderUpdater {
+            address: vec![pinned_address],
+        }));
+    }
+
+    let exit = Arc::new(AtomicBool::new(false));
+    let leader_tpu_service =
+        LeaderTpuService::new(rpc_client, &websocket_url, Protocol::QUIC, exit.clone())
+            .await
+            .map_err(|error| {
+                error!("Failed to create a LeaderTpuService: {error}");
+                LeaderUpdaterError
+            })?;
+    Ok(Box::new(LeaderUpdaterService {
+        leader_tpu_service,
+        exit,
+        lookahead_slots,
+    }))
+}
+
+/// `LeaderUpdaterService` is an implementation of the [`LeaderUpdater`] trait
+/// that dynamically retrieves the current and upcoming leaders by communicating
+/// with the Solana network using [`LeaderTpuService`].
+struct LeaderUpdaterService {
+    leader_tpu_service: LeaderTpuService,
+    exit: Arc<AtomicBool>,
+    /// Number of *estimated* next leaders. If the estimation for the current
+    /// leader is wrong and we have sent to only one leader, we may lose all the
+    /// txs (depends on the forwarding policy).
+    lookahead_slots: u64,
+}
+
+#[async_trait]
+impl LeaderUpdater for LeaderUpdaterService {
+    fn get_leaders(&self) -> Vec<SocketAddr> {
```

Comment:
might be nice to call this something like `get_next_num_lookahead_slots_leaders`. This is a mouthful... so maybe we can come up with something more concise, but the reason I mention this is that when I see the function `get_leaders`, my assumption is we're pulling down all of the leaders for the current epoch.

Another option is to add a function header comment to make it clear we only return the next N leaders based on our estimation of the current leader and the number of lookahead_slots defined when creating the leader updater

Reply:
nit: might be time to stop using java conventions and use `name` and `set_name` for getters and setters :P

Reply:
@bw-solana I agree that `get_leader` might be misleading, changed to `next_num_lookahead_slots_leaders` as proposed.